### PR TITLE
Specify CPU threading

### PR DESCRIPTION
**Pull request by sriharshakandala**
Specify CPU threading

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/devices.jl [hunk 1]
```diff
@@ -12,11 +12,24 @@ abstract type AbstractDevice end
 """
     CPUDevice()
 
-Use the CPU.
+Abstract device type for single-threaded and multi-threaded CPU runs.
 """
-struct CPUDevice <: AbstractDevice end
+abstract type CPUDevice <: AbstractDevice end
```
Comment (charleskawczynski) [inline src/devices.jl:17]:
```suggestion
abstract type AbstractCPUDevice <: AbstractDevice end
```
Should we use the `Abstract` prefix here?

Comment (charleskawczynski) [reply src/devices.jl:17]: I think the name is perhaps better, and it has the benefit of us being able to deprecate `CPUDevice`

File: src/devices.jl [hunk 2]
```diff
@@ -52,19 +67,11 @@ function device()
...
+        return nthreads == 1 ? CPUSingleThreaded() : CPUMultiThreaded()
     end
 end
 
-"""
-    ClimaComms.device(object)
-
-Return the appropriate device for the specified object.
-"""
-device(::Array) = CPUDevice()
```
Comment (charleskawczynski) [inline src/devices.jl:64]: Needing to remove the device makes me wonder if we should instead specialize on the threaded type. For example,

```julia
abstract type AbstractThreading end
struct SingleThreaded <: AbstractThreading end
struct MultiThreaded <: AbstractThreading end

abstract type AbstractDevice end
struct CPUDevice{T<:AbstractThreading} <: AbstractDevice end
struct CUDADevice <: AbstractDevice end
```

Specialization can be more brittle however. What do you think @simonbyrne?

File: src/devices.jl [hunk 3]
```diff
@@ -41,8 +54,10 @@ Behavior can be overridden by setting the `CLIMACOMMS_DEVICE` environment variab
 function device()
     env_var = get(ENV, "CLIMACOMMS_DEVICE", nothing)
     if !isnothing(env_var)
-        if env_var == "CPU"
-            return CPUDevice()
+        if env_var ∈ ("CPU", "CPUSingleThreaded")
+            return CPUSingleThreaded()
```
Comment (simonbyrne) [inline src/devices.jl:58]:
```suggestion
        if env_var == "CPU"
            Threads.nthreads() > 1 ? CPUMultiThreaded() : CPUSingleThreaded()
        elseif env_var = "CPUSingleThreaded"
            return CPUSingleThreaded()
```

File: src/devices.jl [hunk 4]
```diff
@@ -52,19 +67,11 @@ function device()
     if CUDA_jl.functional()
         return CUDADevice()
     else
-        return CPUDevice()
+        nthreads = get(ENV, "JULIA_NUM_THREADS", 1)
```
Comment (simonbyrne) [inline src/devices.jl:70]:
```suggestion
        nthreads = Threads.nthreads()

File: src/devices.jl [hunk 5]
```diff
@@ -52,19 +67,11 @@ function device()
...
+        return nthreads == 1 ? CPUSingleThreaded() : CPUMultiThreaded()
```
Comment (simonbyrne) [inline src/devices.jl:71]:
```suggestion
        return Threads.nthreads() == 1 ? CPUSingleThreaded() : CPUMultiThreaded()
```

File: src/devices.jl (hunk 2 above)
Comment (simonbyrne) [reply src/devices.jl:64]: I don't know. Lets see how this pans out and we can re-evaluate.

File: src/devices.jl (hunk 1 above)
Comment (charleskawczynski) [reply src/devices.jl:17]: I think this conversation was accidentally resolved

Comment (charleskawczynski) [reply src/devices.jl:17]: Ah, the last commit missed the type @sriharshakandala

Comment (sriharshakandala) [reply src/devices.jl]: Sorry, unexpected find/replace error. Its updated now!

**Review by simonbyrne — APPROVED** [review]
Needs a version bump, but otherwise looks good to me.

Before merging, it might be worth making the necessary changes to ClimaCore to check that the interface works.

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
> Needs a version bump, but otherwise looks good to me.
> 
> Before merging, it might be worth making the necessary changes to ClimaCore to check that the interface works.

The ClimaCore tests run correctly with this branch on my local machine. I will go ahead with the merge and bump the version!

**Comment by simonbyrne** [comment]
Can you bump the minor version?

**Comment by sriharshakandala** [comment]
> Can you bump the minor version?

Sure.

**Comment by sriharshakandala** [comment]
https://github.com/CliMA/ClimaComms.jl/pull/53

**Pull request merged**